app/src/org/commcare/android/database/app/AppDatabaseUpgrader.java

```diff
@@ -140,4 +158,80 @@ private boolean upgradeFiveSix(SQLiteDatabase db) {
     private boolean upgradeSixSeven(SQLiteDatabase db) {
         return FixtureSerializationMigration.migrateFixtureDbBytes(db, context);
     }
+
+    /**
+     * Add fields to UserKeyRecord to support PIN auth
+     */
+    private boolean upgradeSevenEight(SQLiteDatabase db) {
+        db.beginTransaction();
+        try {
+            SqlStorage<Persisted> storage = new SqlStorage<Persisted>(
+                    UserKeyRecordV1.STORAGE_KEY,
+                    UserKeyRecordV1.class,
+                    new ConcreteAndroidDbHelper(context, db));
+
+            Vector<UserKeyRecord> migratedRecords = new Vector<>();
+            for (Persisted record : storage) {
+                UserKeyRecordV1 oldUKR = (UserKeyRecordV1)record;
+                UserKeyRecord newUKR = UserKeyRecord.fromOldVersion(oldUKR);
+                newUKR.setID(oldUKR.getID());
+                migratedRecords.add(newUKR);
+            }
+
+            assignActiveRecords(migratedRecords);
+
+            for (UserKeyRecord record : migratedRecords) {
+                storage.write(record);
+            }
+
+            db.setTransactionSuccessful();
+            return true;
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    /**
+     * UserKeyRecordV1 does not have the 'isActive' field (because it is being introduced with
+     * this migration). Go through and mark 1 UKR per username as active, based upon the following
+     * rules:
+     * -If there is only 1 record for a username, mark it as active
+     * -If there are multiple records for a username, mark the one with the latest validTo date as
+     * active
+     */
+    private void assignActiveRecords(Vector<UserKeyRecord> migratedRecords) {
+
+        // First, create a mapping from username --> list of UKRs
+        Map<String, List<UserKeyRecord>> usernamesToRecords = new HashMap<>();
+        for (UserKeyRecord record : migratedRecords) {
+            String username = record.getUsername();
+            List<UserKeyRecord> recordsForUsername = usernamesToRecords.get(username);
+            if (recordsForUsername == null) {
+                recordsForUsername = new ArrayList<>();
+            }
+            recordsForUsername.add(record);
+            usernamesToRecords.put(username, recordsForUsername);
```

Comment:
the `put` call is only needed inside of `recordsForUsername == null`, otherwise the list will be updated regardless.